src/parser/types.rs

```diff
@@ -947,80 +1013,45 @@ impl Stmt {
                 })
             }
 
-            // for loop
+            // for loop, can be either a range or an iterator
             Some(Token {
                 kind: TokenKind::Keyword(Keyword::For),
                 span,
             }) => {
                 tokens.bump(ctx);
 
-                // for i in 0..5 { ... }
+                // for i in arg { ... }
                 //     ^
                 let var = Ident::parse(ctx, tokens)?;
 
-                // for i in 0..5 { ... }
+                // for i in arg { ... }
                 //       ^^
                 tokens.bump_expected(ctx, TokenKind::Keyword(Keyword::In))?;
 
-                // for i in 0..5 { ... }
+                // for i in arg { ... }
                 //          ^
-                let (start, start_span) = match tokens.bump(ctx) {
-                    Some(Token {
-                        kind: TokenKind::BigUInt(n),
-                        span,
-                    }) => {
-                        let start: u32 = n
-                            .try_into()
-                            .map_err(|_e| ctx.error(ErrorKind::InvalidRangeSize, span))?;
-                        (start, span)
-                    }
-                    _ => {
-                        return Err(ctx.error(
-                            ErrorKind::ExpectedToken(TokenKind::BigUInt(
-                                num_bigint::BigUint::zero(),
-                            )),
-                            ctx.last_span(),
-                        ))
-                    }
-                };
-
-                // for i in 0..5 { ... }
-                //           ^^
-                tokens.bump_expected(ctx, TokenKind::DoubleDot)?;
-
-                // for i in 0..5 { ... }
-                //             ^
-                let (end, end_span) = match tokens.bump(ctx) {
+                // here we decide if this for loop is over a range or is an iterator,
```

Comment:
This is not always the case once the generic feature is introduced. For example:

```rust
fn range(const START: Field, const END: Field) {
    for ii in START..END { // START / END are const generic variables, which are not in numeric form in the syntax.
        ...
    }
}
```

So I think the safer way to detect whether it is range or iterator can be based on the presence of `..`, which indicates an range.